### PR TITLE
Properly document the local and global locality attributes.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** documentation

Following the last Coq Call (https://github.com/coq/coq/wiki/Coq-Call-2021-02-03), this PR properly documents the semantics of the locality attribute `global` and `local`. It also fixes incorrect documentation about the default locality of `Set` / `Unset`.

### CODE REVIEW

File: doc/sphinx/language/core/basic.rst [hunk 1]
```diff
@@ -527,20 +527,20 @@ The :cmd:`Set` and :cmd:`Unset` commands support the :attr:`local`,
...
+* :attr:`export` (or alternatively, the ``Export`` prefix): the
+  original setting is *not* restored at the end of the current section
+  but it is restored at the end of the current module. In addition,
+  the setting persists when :cmd:`Import`\ing the module in which it
+  is set.
 * :attr:`global` (or alternatively, the ``Global`` prefix): the
   original setting is *not* restored at the end of the current module
   or section.  In addition, if the value is set in a file, then
-  :cmd:`Require`\-ing the file sets the option.
+  :cmd:`Require`\-ing (even transitively) the file sets the option.
```
Comment (jfehrle) [inline doc/sphinx/language/core/basic.rst:543]:
```suggestion
  :cmd:`Require`\-ing the file (even transitively) sets the option.
```

The language in this section is confusing:
- we have "settings", which are flags and options, both of which can be "set" and "unset".
- "sets the option" doesn't seem to include flags
- if you unset a flag, can the "false" be exported?  If so, "if the value is set in a file" is misleading

I think we need a word that means "option or flag" other than "setting".  Or just use that phrase
repeatedly for clarity.  Maybe use "value" in place of some uses of "setting".

File: doc/sphinx/language/core/basic.rst [hunk 2]
```diff
@@ -527,20 +527,20 @@ The :cmd:`Set` and :cmd:`Unset` commands support the :attr:`local`,
...
-* :attr:`export` (or alternatively, the ``Export`` prefix): similar to
-  :attr:`local`, the original value of the setting is restored at the
-  end of the current module or section.  In addition, if the value is
-  set in a module, then :cmd:`Import`\-ing the module sets the option
-  or flag.
+* :attr:`export` (or alternatively, the ``Export`` prefix): the
+  original setting is *not* restored at the end of the current section
+  but it is restored at the end of the current module. In addition,
+  the setting persists when :cmd:`Import`\ing the module in which it
+  is set.
```
Comment (jfehrle) [inline doc/sphinx/language/core/basic.rst:539]:
```suggestion
  exported settings in :cmd:`Import`\ed modules override the current value
```

File: doc/sphinx/language/core/modules.rst [hunk 3]
```diff
@@ -1010,3 +1010,66 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+Controlling the scope of commands with locality attributes
+----------------------------------------------------------
+
+Many commands produce an effect that will persist in a given scope,
+which can typically be the one of the section or of the module in
+which the command was called. Locality :term:`attributes <attribute>`
+can be used to change the scope of this effect. Note that some
+commands, like :cmd:`Arguments` do not have the same default locality
+when they are used inside or outside sections.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1024]:
```suggestion
Many commands have effects that apply only within a specific scope,
typically the section or the module in
which the command was called. Locality :term:`attributes <attribute>`
can alter the scope of the effect. Note that some
commands, like :cmd:`Arguments` do not have the same default locality
when they are used inside or outside sections.
```

"typically"?  what other scopes are there other than sections and modules?
"Note that .." should be under `attr:: local`.  In this paragraph you could say
that a few commands don't follow the general pattern--that alerts the reader
that quirks lie ahead.

File: doc/sphinx/language/core/modules.rst [hunk 4]
```diff
@@ -1010,3 +1010,66 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+.. attr:: local
+
+   In most cases, the :attr:`local` attribute limits the effect of the
+   command to the current scope (section or module).
+
+   The only exception is when applied to :cmd:`Definition`,
+   :cmd:`Theorem` or variants. For these commands, the effect of the
+   :attr:`local` attribute will be that the defined object will not be
+   available through a shorter name when importing the module in which
+   it is defined.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1035]:
```suggestion
   The only exception is when :attr:`local` is applied to :cmd:`Definition`,
   :cmd:`Theorem` or their variants. Importing modules containing
   such commands makes the defined objects available only through
   their fully-qualified names rather than their unqualified names.
```

File: doc/sphinx/language/core/modules.rst [hunk 5]
```diff
@@ -1010,3 +1010,66 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+   There is an alternative syntax for the :attr:`local` attribute
+   using the ``Local`` prefix (see :n:`@legacy_attr`).
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1038]:
```suggestion
   The``Local`` prefix is an alternative syntax for the :attr:`local` attribute
   (see :n:`@legacy_attr`).
```

File: doc/sphinx/language/core/modules.rst [hunk 6]
```diff
@@ -1010,3 +1010,66 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+   .. note::
+
+      For some commands, this is the only supported locality within
+      sections (e.g., for :cmd:`Notation`, :cmd:`Ltac` and :ref:`Hint
+      <creating_hints>` commands).
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1044]: "only supported locality"??  Not clear what is meant.  If you mean that Notation, etc don't allow global or export, say that more directly, and maybe under the description of those attributes

File: doc/sphinx/language/core/modules.rst [hunk 7]
```diff
@@ -1010,3 +1010,66 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+.. attr:: export
+
+   The :attr:`export` attribute makes the effect of the command
+   persist when the section is closed and when the module containing
+   the command is imported.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1050]:
```suggestion
   The :attr:`export` attribute makes the effect of the command
   persist when the section is closed and applies the effect when the module containing
   the command is imported.
```

"persist" suggests that the effect is already present, not quite right for importing.

File: doc/sphinx/language/core/modules.rst [hunk 8]
```diff
@@ -1010,3 +1010,66 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+   This attribute is supported in particular by the :ref:`Hint
+   <creating_hints>`, :cmd:`Set` and :cmd:`Unset` commands, although
+   for :ref:`Hint <creating_hints>` commands, it is not supported
+   inside sections.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1055]:
```suggestion
   This attribute is supported in particular by the :ref:`Hint
   <creating_hints>`, :cmd:`Set` and :cmd:`Unset` commands, although
   for :ref:`Hint <creating_hints>` commands, it is not supported
   inside sections.
```

is "in particular" needed?  There are others that are not important.

File: doc/sphinx/language/core/modules.rst [hunk 9]
```diff
@@ -1010,3 +1010,66 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+   In most cases, the :attr:`global` attribute makes the effect of the
+   command persist even when the current section or module is closed
+   and in other files as soon as the one containing the command is
+   (possibly transitively) loaded.
+
+   For a few commands (like :cmd:`Notation` and :cmd:`Ltac`), this
+   attribute behaves as in the description of :attr:`export`.
+
+   There is an alternative syntax for the :attr:`global` attribute
+   using the ``Global`` prefix (see :n:`@legacy_attr`).
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1068]:
```suggestion
   The ``Global`` prefix is an alternative syntax for the :attr:`global` attribute
   (see :n:`@legacy_attr`).
```

File: doc/sphinx/language/core/modules.rst [hunk 10]
```diff
@@ -1010,3 +1010,66 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+.. attr:: global
+
+   In most cases, the :attr:`global` attribute makes the effect of the
+   command persist even when the current section or module is closed
+   and in other files as soon as the one containing the command is
+   (possibly transitively) loaded.
+
+   For a few commands (like :cmd:`Notation` and :cmd:`Ltac`), this
+   attribute behaves as in the description of :attr:`export`.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1065]:
```suggestion
   In most cases, the :attr:`global` attribute makes the effect of the
   command persist even when the current section or module is closed.
   Loading the module (possibly transitively) applies the effect of the command.

   For a few commands (like :cmd:`Notation` and :cmd:`Ltac`), this
   attribute behaves like :attr:`export`.
```

Maybe leave out "In most cases," and just have a label "Exceptions" here and in a couple other spots?
Not clear to me whether the exceptions are well thought out or more like historical accidents.

File: doc/sphinx/language/core/modules.rst [hunk 11]
```diff
@@ -1010,3 +1010,66 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+   There is an alternative syntax for the :attr:`global` attribute
+   using the ``Global`` prefix (see :n:`@legacy_attr`).
+
+   .. warning::
+
+      Using the :attr:`global` locality is strongly discouraged
+      because the transitive nature of file loading gives little
+      control to the user. Prefer the use of the :attr:`export`
+      locality attribute when it is supported.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1075]:
```suggestion
      We strongly discourage using :attr:`global` in commands
      because the transitive nature of file loading gives the user little
      control. We recommend using the :attr:`export`
      locality attribute where it is supported.
```

**Review by jfehrle — COMMENTED** [review]
More comments than I expected to make.

Should someone else review it for content since I'm not the expert on this?

File: doc/sphinx/language/core/basic.rst [hunk 12]
```diff
@@ -527,20 +527,20 @@ The :cmd:`Set` and :cmd:`Unset` commands support the :attr:`local`,
 :attr:`global` and :attr:`export` locality attributes:
 
 * no attribute: the original setting is *not* restored at the end of
```
[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/sphinx/language/core/basic.rst:529]: BTW, do we like the not restoring at the end of a section behavior or is it judged too disruptive to try to change it to something else?

File: doc/sphinx/language/core/basic.rst [hunk 13]
```diff
@@ -527,20 +527,20 @@ The :cmd:`Set` and :cmd:`Unset` commands support the :attr:`local`,
...
-  the current module or section.
+  the current section but it is restored at the end of the current module.
 * :attr:`local` (or alternatively, the ``Local`` prefix): the setting
```
[the remark this answers is not in the document]
Comment (Zimmi48) [reply doc/sphinx/language/core/basic.rst]: I'd like to keep the current order for now, but I've changed the description of `export` to factorize more.

File: doc/sphinx/language/core/basic.rst (hunk 1 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/basic.rst]: >if you unset a flag, can the "false" be exported?

Yes.

Note that "setting" also includes table. Otherwise, option would be a generic term (flags are boolean options).

I will try to improve the wording.

File: doc/sphinx/language/core/modules.rst (hunk 3 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst]: > "typically"? what other scopes are there other than sections and modules?

Some commands have effects which persist module boundaries by default.

I've added an alert. Let me know what you think.

File: doc/sphinx/language/core/modules.rst (hunk 6 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst]: These commands come with a locality (either the default one or the one set by the attribute). I think this explanation belongs here. It wouldn't make sense to repeat it for every other locality attribute, especially as we extend the list.

File: doc/sphinx/language/core/modules.rst (hunk 8 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst:1062]: The goal was to prevent the documentation to become incorrect as the list of commands supporting the attribute would expand.

File: doc/sphinx/language/core/modules.rst (hunk 10 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst]: The exceptions are the result of history, but we should get rid of them eventually.

**Comment by Zimmi48** [comment]
> More comments than I expected to make.

Thanks a lot for these. The text is much improved now.

> Should someone else review it for content since I'm not the expert on this?

Yes, I would appreciate if @ppedrot, @gares or @SkySkimmer could review this PR as well.

File: doc/sphinx/language/core/modules.rst (hunk 10 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst]: See #13153 for the exception to the default semantics of `local` and #13833 for the exceptions to the default semantics of `global`.

Comment (jfehrle) [reply doc/sphinx/language/core/modules.rst]: While I don't want to become expert on all the variations, it would be nice to get to a point where there is a 1-1 relationship between attributes and behaviors.  For example, if superglobal behavior remains available for some commands, we might create a `superglobal` attribute to distinguish it from regular global behavior.  Such commands could also give an error message like "'global' is not supported, did you mean 'superglobal'"?

File: doc/sphinx/language/core/modules.rst (hunk 8 above)
Comment (jfehrle) [reply doc/sphinx/language/core/modules.rst:1062]: You could say "Commands supporting this attribute include ...".  "In particular" implies that the listed items are somehow more important that those that are not listed.

File: doc/sphinx/language/core/basic.rst (hunk 1 above)
Comment (jfehrle) [reply doc/sphinx/language/core/basic.rst]: Yes, "settings" includes tables, but the `Set` and `Unset` commands don't apply to tables, which use `Add` and `Remove` instead.  That's a little confusing.

File: doc/sphinx/language/core/basic.rst [hunk 14]
```diff
@@ -523,24 +523,24 @@ they appear after a boldface label.  They are listed in the
...
-  the current module or section.
-* :attr:`local` (or alternatively, the ``Local`` prefix): the setting
+The :cmd:`Set` and :cmd:`Unset` commands support the mutually
+exclusive :attr:`local`, :attr:`export` and :attr:`global` locality
+attributes:
+
+* if no attribute is specified: the original value of the setting is
+  restored at the end of the current module but it is *not* restored
+  at the end of the current section.
+* :attr:`local` (or the ``Local`` prefix): the setting
```
Comment (jfehrle) [inline doc/sphinx/language/core/basic.rst:533]:
```suggestion
* :attr:`local` (or the ``Local`` :n:`@legacy_attr`): the setting
```

and similar changes elsewhere

File: doc/sphinx/language/core/basic.rst [hunk 15]
```diff
@@ -523,24 +523,24 @@ they appear after a boldface label.  They are listed in the
...
-  end of the current module or section.  In addition, if the value is
-  set in a module, then :cmd:`Import`\-ing the module sets the option
-  or flag.
-* :attr:`global` (or alternatively, the ``Global`` prefix): the
-  original setting is *not* restored at the end of the current module
-  or section.  In addition, if the value is set in a file, then
-  :cmd:`Require`\-ing the file sets the option.
+* :attr:`export` (or the ``Export`` prefix): same as no attribute, but
+  in addition, exported settings in :cmd:`Import`\ed modules override
+  the current value.
```
Comment (jfehrle) [inline doc/sphinx/language/core/basic.rst:539]:
```suggestion
* :attr:`export` (or the ``Export`` prefix): similar to no attribute, but
  settings exported from :cmd:`Import`\ed modules override
  the current value.
```

Or maybe "same as no attribute, except ...".  "In addition" is a weak way to convey the contradiction between the two clauses.  IMHO.

File: doc/sphinx/language/core/basic.rst [hunk 16]
```diff
@@ -523,24 +523,24 @@ they appear after a boldface label.  They are listed in the
...
+* :attr:`global` (or the ``Global`` prefix): the original value of the
+  setting is *not* restored at the end of the current module or
+  section.  In addition, global settings in a file override the
+  current value when the file is :cmd:`Require`\d (even transitively).
```
Comment (jfehrle) [inline doc/sphinx/language/core/basic.rst:543]: I notice you refer to files here and to modules above (line 538).  Is there a difference?  Maybe worth mentioning/repeating the distinction here (if there is one)?

File: doc/sphinx/language/core/modules.rst [hunk 17]
```diff
@@ -1010,3 +1010,75 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+Controlling the scope of commands with locality attributes
+----------------------------------------------------------
+
+Many commands have effects that apply only within a specific scope,
+typically the section or the module in which the command was
+called. Locality :term:`attributes <attribute>` can alter the scope of
+the effect. Below, we give the semantics of each locality attribute
+but also alert that a few commands give a different meaning to the
+:attr:`local` and :attr:`global` attributes.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1024]:
```suggestion
while noting a few exceptional commands in which 
:attr:`local` and :attr:`global` attributes are interpreted differently.
```

File: doc/sphinx/language/core/modules.rst [hunk 18]
```diff
@@ -1010,3 +1010,75 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+.. attr:: local
+
+   The :attr:`local` attribute limits the effect of the command to the
+   current scope (section or module).
+
+   The ``Local`` prefix is an alternative syntax for the :attr:`local`
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1031]:
```suggestion
   The ``Local`` :n:`@legacy_attr` is an alternative syntax for the :attr:`local`
```

or "legacy attribute" if that can be made to hyperlink

File: doc/sphinx/language/core/modules.rst [hunk 19]
```diff
@@ -1010,3 +1010,75 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+   attribute (see :n:`@legacy_attr`).
+
+   .. note::
+
+      - For some commands, this is the only supported locality within
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1036]:
```suggestion
      - For some commands, this is the only locality supported within
```

File: doc/sphinx/language/core/modules.rst [hunk 20]
```diff
@@ -1010,3 +1010,75 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+        sections (e.g., for :cmd:`Notation`, :cmd:`Ltac` and
+        :ref:`Hint <creating_hints>` commands).
+
+      - For some other commands, this is the default locality within
+        sections even though other locality attributes are supported
+        as well (e.g., it is the case for the :cmd:`Arguments`
+        command).
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1043]:
```suggestion
      - For other commands, this is the default locality within
        sections even though other locality attributes are supported
        as well (e.g., for the :cmd:`Arguments` command).
```

Of the two cases, "the only supported locality" and "the default locality", which is more more common?  If there's a definite preference for one or the other, maybe say "For _most_ commands"

File: doc/sphinx/language/core/modules.rst [hunk 21]
```diff
@@ -1010,3 +1010,75 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+   .. warning::
+
+      **Exception:** when :attr:`local` is applied to
+      :cmd:`Definition`, :cmd:`Theorem` or their variants, its
+      semantics is different: it makes the defined objects available
+      only through their fully-qualified names rather than their
+      unqualified names after an :cmd:`Import`.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1051]:
```suggestion
      :cmd:`Definition`, :cmd:`Theorem` or their variants, its
      semantics are different: when these objects are :cmd:`Import`\ed,
      they must be referenced with their fully-qualified names rather than their
      unqualified names.
```

File: doc/sphinx/language/core/modules.rst [hunk 22]
```diff
@@ -1010,3 +1010,75 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+.. attr:: export
+
+   The :attr:`export` attribute makes the effect of the command
+   persist when the section is closed and applies the effect when the
+   module containing the command is imported.
+
+   This attribute is supported in particular by the :ref:`Hint
+   <creating_hints>`, :cmd:`Set` and :cmd:`Unset` commands, although
+   for :ref:`Hint <creating_hints>` commands, it is not supported
+   inside sections.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1062]:
```suggestion
   Commands supporting this attribute include :cmd:`Set`, :cmd:`Unset` and
   the :ref:`Hint <creating_hints>` commands, although the latter doesn't support it
   within sections.
```

File: doc/sphinx/language/core/modules.rst [hunk 23]
```diff
@@ -1010,3 +1010,75 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+
+      **Exception:** for a few commands (like :cmd:`Notation` and
+      :cmd:`Ltac`), this attribute behaves like :attr:`export`.
+
+   .. warning::
+
+      We strongly discourage using the :attr:`global` locality
+      attribute because the transitive nature of file loading gives
+      the user little control. We recommend using the :attr:`export`
+      locality attribute where it is supported.
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:1084]: The idea is to eventually phase out "global"?

**Review by jfehrle — COMMENTED** [review]
Some more wording suggestions, but already looks good.

**Comment by gares** [comment]
In the initial part of the patch, I'm not sure it is correct to say `export` is like "nothing + exporting". For some commands, nothing means export.
IMO we should try to finish/update this https://https://github.com/coq/coq/wiki/Default-Locality-Behavior-of-Commands and then have a complete list of exceptions.

File: doc/sphinx/language/core/modules.rst [hunk 24]
```diff
@@ -1010,3 +1010,75 @@ subdirectories of path). See the command :cmd:`Declare ML Module` in
...
+   the file containing the command (possibly transitively) applies the
+   effect of the command.
+
+   The ``Global`` prefix is an alternative syntax for the
+   :attr:`global` attribute (see :n:`@legacy_attr`).
+
+   .. warning::
+
+      **Exception:** for a few commands (like :cmd:`Notation` and
+      :cmd:`Ltac`), this attribute behaves like :attr:`export`.
```
Comment (SkySkimmer) [inline doc/sphinx/language/core/modules.rst:1077]: This sounds like they also support `export`, but they don't.

File: doc/sphinx/language/core/basic.rst [hunk 25]
```diff
@@ -523,24 +523,24 @@ they appear after a boldface label.  They are listed in the
...
 
-The :cmd:`Set` and :cmd:`Unset` commands support the :attr:`local`,
-:attr:`global` and :attr:`export` locality attributes:
-
-* no attribute: the original setting is *not* restored at the end of
-  the current module or section.
-* :attr:`local` (or alternatively, the ``Local`` prefix): the setting
+The :cmd:`Set` and :cmd:`Unset` commands support the mutually
+exclusive :attr:`local`, :attr:`export` and :attr:`global` locality
+attributes:
```
Comment (SkySkimmer) [inline doc/sphinx/language/core/basic.rst:528]: Do we need the full description of effects here? We could just reference to `_controlling-locality-of-commands` with a note that Set also has special alias syntax `Export`

**Comment by Zimmi48** [comment]
> In the initial part of the patch, I'm not sure it is correct to say `export` is like "nothing + exporting". For some commands, nothing means export.

This is specific to locality attributes applied to setting commands (`Set` / `Unset`). See also #13832 and in particular my response to @SkySkimmer's comment and the example it contains.

File: doc/sphinx/language/core/basic.rst [hunk 26]
```diff
@@ -523,24 +523,24 @@ they appear after a boldface label.  They are listed in the
...
+
+* if no attribute is specified: the original value of the setting is
+  restored at the end of the current module but it is *not* restored
+  at the end of the current section.
```
[the remark this answers is not in the document]
Comment (Zimmi48) [reply doc/sphinx/language/core/basic.rst:532]: No, see the example I added to #13832.

File: doc/sphinx/language/core/modules.rst (hunk 24 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst:1077]: In the previous version of the patch I wrote "behaves as in the description of `export`" but it seemed a bit wordy.

File: doc/sphinx/language/core/basic.rst (hunk 25 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/basic.rst:528]: Indeed, now that we have a proper documentation of all locality attributes, we do not need to repeat it here (except for the weird default locality). Adding an explicit attribute for the default locality (as proposed in #13832) would even solve this remaining issue.

File: doc/sphinx/language/core/modules.rst (hunk 23 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst:1084]: I cannot say for sure if it will be ever deprecated but yes.

File: doc/sphinx/language/core/modules.rst (hunk 21 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst:1051]: I've applied the "is" -> "are" change but not the rest because I don't want to lose the idea of the effect of `local` that is carried by "it makes".

File: doc/sphinx/language/core/modules.rst (hunk 20 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst:1043]: The two cases are not the only options (which seems implied by the "For other commands"). I've changed it into "For some" for more symmetry. There's no definite preference and the first case (only supported locality) is a limitation rather than a feature.

File: doc/sphinx/language/core/modules.rst (hunk 18 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst:1031]: I prefer to keep as is.

File: doc/sphinx/language/core/basic.rst (hunk 16 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/basic.rst:543]: Yes, you can `Import` a module but only `Require` a file. Because a file is also a module, you can `Require Import` it. This whole chapter needs to be improved and consolidated anyway to make this fact clearer.

File: doc/sphinx/language/core/basic.rst (hunk 14 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/basic.rst:533]: I don't think it improves readability. Furthermore, there is the special case of `Export` which isn't a proper legacy attribute (as we've previously discussed) and which would break symmetry.

File: doc/sphinx/language/core/basic.rst (hunk 1 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/basic.rst]: Indeed. And you just made me dig what was the behavior for `Add` and `Remove` and I was surprised to discover it is actually different (the default is `local` and `export` is not supported):

```coq
Section section.

  Add Search Blacklist "foo".

  Test Search Blacklist for "foo".
  (* Search blacklist does include foo *)

End section.

Test Search Blacklist for "foo".
(* Search blacklist does not include foo *)
```

File: doc/sphinx/language/core/modules.rst (hunk 10 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst]: The most common semantics of `global` is the superglobal one, so the idea is rather to eventually make sure this is the only semantics behind `global` (and to reduce the number of commands which support it).

File: doc/sphinx/language/core/basic.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (Zimmi48) [reply doc/sphinx/language/core/basic.rst]: Now the descriptions of standard locality attributes `local`, `export` and `global` are not included there anymore.

File: doc/sphinx/language/core/basic.rst (hunk 1 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/basic.rst]: I've changed the text to use the wording "flag or option" instead of "setting".

File: doc/sphinx/language/core/basic.rst (hunk 25 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/basic.rst]: I've applied your suggestion. Now this part of the text doesn't repeat the semantics of `local`, `export` and `global`.

**Comment by Zimmi48** [comment]
> IMO we should try to finish/update this https://github.com/coq/coq/wiki/Default-Locality-Behavior-of-Commands and then have a complete list of exceptions.

IMHO it would be almost as efficient to just implement #13153 and #13833 to get rid of the exceptions. The linked page only talks about the default locality BTW, not the semantics given to `Global`.

File: doc/sphinx/language/core/modules.rst (hunk 6 above)
Comment (jfehrle) [reply doc/sphinx/language/core/modules.rst]: Perhaps "locality" should be "locality option" everywhere since "locality" is not a defined term.  The dictionary definition is "the position or site of something" though in CS it's used a bit more broadly.

**Review by jfehrle — APPROVED** [review]
The wording looks fine.

File: doc/sphinx/language/core/modules.rst (hunk 6 above)
Comment (Zimmi48) [reply doc/sphinx/language/core/modules.rst]: The more appropriate term should really be "scope", but my goal in this PR is not to revise the entire thing and locality is the term that has been used so far. In most places, I write "locality attribute" which is a tad more explicit.

**Comment by Zimmi48** [comment]
From my point of view, this PR is now ready to merge.

**Comment by jfehrle** [comment]
Would be nice if @SkySkimmer of @gares would approve the content, but perhaps that's not necessary.

**Review by gares — APPROVED** [review]

**Comment by jfehrle** [comment]
Thanks for the PR!
@coqbot: merge now

**Comment by jfehrle** [comment]
@coqbot: merge now, dammit!
